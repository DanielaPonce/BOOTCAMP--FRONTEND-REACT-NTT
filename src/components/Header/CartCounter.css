.cart-icon {
	cursor: pointer;
	position: relative;
	width: 30px;
	height: 30px;
}

.cart-icon .cart-count {
	position: absolute;
	top: -8px;
	right: -8px;
	background-color: var(--error-color);
	color: var(--secondary-color);
	border-radius: 50%;
	padding: 0.2rem 0.5rem;
	font-size: 0.8rem;
}

@media only screen and (max-width: 739px) {
	.cart-icon {
		position: absolute;
		top: 1.5rem;
		right: 2rem;
	}

}