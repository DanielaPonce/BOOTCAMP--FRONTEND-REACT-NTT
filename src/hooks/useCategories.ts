import { useEffect, useState } from 'react';
import { fetchAllCategories } from '../api/category.api';
import { Category } from '../models/category.types';

export const useCategories = () => {
	const [categories, setCategories] = useState<Category[]>([]);

	useEffect(() => {
		fetchAllCategories().then((data) => setCategories(data));
	}, []);

	return { categories };
};
