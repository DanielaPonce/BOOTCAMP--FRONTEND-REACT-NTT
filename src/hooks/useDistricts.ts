import { useEffect, useState } from 'react';

interface District {
	id: number;
	name: string;
}

export const useDistricts = () => {
	const [districts, setDistricts] = useState<District[]>([]);

	useEffect(() => {
		const fetchDistricts = async () => {
			const response = await fetch('/src/assets/mock/districts.json');
			const data: District[] = await response.json();
			setDistricts(data);
		};

		fetchDistricts();
	}, []);

	return districts;
};
