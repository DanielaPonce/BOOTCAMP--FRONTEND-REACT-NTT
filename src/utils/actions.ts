export enum Actions {
	ADD_TO_CART_TYPE = 'ADD_TO_CART',
	REMOVE_FROM_CART_TYPE = 'REMOVE_FROM_CART',
	UPDATE_QUANTITY_TYPE = 'UPDATE_QUANTITY',
	CLEAR_CART_TYPE = 'CLEAR_CART'
}
