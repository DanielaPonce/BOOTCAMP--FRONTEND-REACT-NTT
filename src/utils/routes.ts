export enum RoutesConstants {
	home = '/',
	summary = 'summary'
}
