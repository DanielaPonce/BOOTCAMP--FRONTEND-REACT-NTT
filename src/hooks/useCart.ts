import { useEffect } from 'react';
import { useCartContext } from '../context/CartContext';
import { Product } from '../models/product.types';
import { Actions } from '../utils/actions';
import { LocalStorageConstants } from '../utils/localStorageConstants';

export const useCart = () => {
	const { state, dispatch } = useCartContext();

	useEffect(() => {
		localStorage.setItem(
			LocalStorageConstants.CartKey,
			JSON.stringify(state)
		);
	}, [state]);

	const addToCart = (product: Product) =>
		dispatch({ type: Actions.ADD_TO_CART_TYPE, payload: product });

	const removeFromCart = (id: number) =>
		dispatch({ type: Actions.REMOVE_FROM_CART_TYPE, payload: id });

	const updateQuantity = (id: number, quantity: number) =>
		dispatch({
			type: Actions.UPDATE_QUANTITY_TYPE,
			payload: { id, quantity }
		});

	const clearCart = () => dispatch({ type: Actions.CLEAR_CART_TYPE });

	return {
		cart: state.cart,
		total: state.total,
		itemCount: state.itemCount,
		addToCart,
		removeFromCart,
		updateQuantity,
		clearCart
	};
};
