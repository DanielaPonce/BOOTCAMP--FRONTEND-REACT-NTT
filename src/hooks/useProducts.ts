import { useEffect, useState } from 'react';
import { fetchAllProducts, fetchProductsByCategory } from '../api/product.api';
import { Product } from '../models/product.types';
import { Constants } from '../utils/constants';

export const useProducts = (category: string, search: string) => {
	const [allProducts, setAllProducts] = useState<Product[]>([]);
	const [filteredProducts, setFilteredProducts] = useState<Product[]>([]);
	const [isLoading, setIsLoading] = useState(true);
	const [currentPage, setCurrentPage] = useState(1);

	const itemsPerPage = 8;

	const loadProducts = async () => {
		setIsLoading(true);
		try {
			let products: Product[] = [];
			if (category === Constants.defaultSelectValue) {
				products = await fetchAllProducts();
			} else {
				products = await fetchProductsByCategory(category);
			}
			setAllProducts(products);
			setFilteredProducts(products);
			setCurrentPage(1);
		} catch (error) {
			console.log('Error al cargar los productos:', error);
		} finally {
			setIsLoading(false);
		}
	};

	useEffect(() => {
		loadProducts();
	}, [category]);

	useEffect(() => {
		const filtered = allProducts.filter((product) =>
			product.title.toLowerCase().includes(search.toLowerCase())
		);
		setFilteredProducts(filtered);
		setCurrentPage(1);
	}, [search, allProducts]);

	const paginatedProducts = filteredProducts.slice(
		(currentPage - 1) * itemsPerPage,
		currentPage * itemsPerPage
	);

	const totalPages = Math.ceil(filteredProducts.length / itemsPerPage);

	return {
		products: paginatedProducts,
		isLoading,
		totalPages,
		currentPage,
		setCurrentPage,
		showPagination: totalPages !== 1
	};
};
