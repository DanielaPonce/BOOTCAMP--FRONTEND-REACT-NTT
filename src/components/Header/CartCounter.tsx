import { FC } from 'react';
import { useCart } from '../../hooks/useCart';
import { Link } from 'react-router';
import { RoutesConstants } from '../../utils/routes';

const CartCounter: FC = () => {
	const { itemCount } = useCart();

	return (
		<div className="cart-icon">
			<Link to={RoutesConstants.summary}>
				<img src="/src/assets/icons/cart-icon.svg" alt="cart icon" />
			</Link>

			{itemCount > 0 ? (
				<span id="cardCounter" className="cart-count">
					{itemCount}
				</span>
			) : (
				<></>
			)}
		</div>
	);
};

export default CartCounter;
