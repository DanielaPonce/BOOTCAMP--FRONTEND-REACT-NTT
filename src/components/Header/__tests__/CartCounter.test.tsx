import { jest } from '@jest/globals';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { useCart } from '../../../hooks/useCart';
import { RoutesConstants } from '../../../utils/routes';
import CartCounter from '../CartCounter';

jest.mock('../../../hooks/useCart');

describe('CartCounter Component', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	test('renders correctly', () => {
		(useCart as jest.Mock).mockReturnValue({ itemCount: 0 });

		render(
			<MemoryRouter>
				<CartCounter />
			</MemoryRouter>
		);

		const cartIcon = screen.getByRole('img', { name: /cart icon/i });
		expect(cartIcon).toBeInTheDocument();
		expect(cartIcon).toHaveAttribute(
			'src',
			'/src/assets/icons/cart-icon.svg'
		);

		const link = screen.getByRole('link');
		expect(link).toHaveAttribute('href', RoutesConstants.Summary);
	});

	test('shows the item count when itemCount > 0', () => {
		(useCart as jest.Mock).mockReturnValue({ itemCount: 5 });

		render(
			<MemoryRouter>
				<CartCounter />
			</MemoryRouter>
		);

		const cartCount = screen.getByText('5');
		expect(cartCount).toBeInTheDocument();
		expect(cartCount).toHaveClass('cart-count');
	});

	test('hides the item count when itemCount is 0', () => {
		(useCart as jest.Mock).mockReturnValue({ itemCount: 0 });

		render(
			<MemoryRouter>
				<CartCounter />
			</MemoryRouter>
		);

		const cartCount = screen.queryByText('0');
		expect(cartCount).toBeNull();
	});

	test('matches snapshot', () => {
		(useCart as jest.Mock).mockReturnValue({ itemCount: 3 });

		const { asFragment } = render(
			<MemoryRouter>
				<CartCounter />
			</MemoryRouter>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
