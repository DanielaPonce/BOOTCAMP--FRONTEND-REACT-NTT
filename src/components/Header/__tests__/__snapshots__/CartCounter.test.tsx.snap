// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartCounter Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="cart-icon"
  >
    <a
      data-discover="true"
      href="/summary"
    >
      <img
        alt="cart icon"
        src="/src/assets/icons/cart-icon.svg"
      />
    </a>
    <span
      class="cart-count"
      id="cardCounter"
    >
      3
    </span>
  </div>
</DocumentFragment>
`;
